### PR TITLE
Add allOf support

### LINKED ISSUE #52: Support for joined schemas with `anyOf`, `allOf`, etc?
I'm trying to combine two JSON schemas together like so:

```
let schema = {
  type: 'object',
  anyOf: [
    schema1,
    schema2,
  ],
}
```

Doesn't look like this is supported, or did I miss it? Any plans to add support?

### LINKED ISSUE #1237: type is undefined but have allOf
Hi, I have a very long jsonschema which I want see as form.
Using the live demo, have this error:

image

doing some test using the live demo, I found that if type is not declared, throw the error, its possible use allOf without type?

### LINKED ISSUE #1378: minItems does not override a ref to an array
### Prerequisites

- [x] I have read the [documentation](https://react-jsonschema-form.readthedocs.io/);
- [x] In the case of a bug report, I understand that providing a [SSCCE](http://sscce.org/) example is tremendously useful to the maintainers.
- [x] Ideally, I'm providing a [sample JSFiddle](https://jsfiddle.net/n1k0/f2y3fq7L/6/) or a [shared playground link](https://mozilla-services.github.io/react-jsonschema-form/) demonstrating the issue.

### Description

minItems does not override a ref to an array.

### Steps to Reproduce

Try [this schema](https://mozilla-services.github.io/react-jsonschema-form/#eyJmb3JtRGF0YSI6eyJwYXJ0aWNpcGFudHMiOlsiYiJdfSwic2NoZW1hIjp7ImRlZmluaXRpb25zIjp7InBhcnRpY2lwYW50cyI6eyJ0eXBlIjoiYXJyYXkiLCJpdGVtcyI6eyJ0eXBlIjoic3RyaW5nIn19fSwidHlwZSI6Im9iamVjdCIsInByb3BlcnRpZXMiOnsicGFydGljaXBhbnRzIjp7IiRyZWYiOiIjL2RlZmluaXRpb25zL3BhcnRpY2lwYW50cyIsIm1pbkl0ZW1zIjozfX19LCJ1aVNjaGVtYSI6e319).

#### Expected behavior

minItems should be 3, so it should give an error.

#### Actual behavior

No error is given; thus, minItems has not been respected.

Note, though, that the expected behavior is shown when minItems is passed in directly inside the ref object, at [this link](https://mozilla-services.github.io/react-jsonschema-form/#eyJmb3JtRGF0YSI6eyJwYXJ0aWNpcGFudHMiOlsiYiJdfSwic2NoZW1hIjp7ImRlZmluaXRpb25zIjp7InBhcnRpY2lwYW50cyI6eyJ0eXBlIjoiYXJyYXkiLCJpdGVtcyI6eyJ0eXBlIjoic3RyaW5nIn0sIm1pbkl0ZW1zIjozfX0sInR5cGUiOiJvYmplY3QiLCJwcm9wZXJ0aWVzIjp7InBhcnRpY2lwYW50cyI6eyIkcmVmIjoiIy9kZWZpbml0aW9ucy9wYXJ0aWNpcGFudHMifX19LCJ1aVNjaGVtYSI6e319).

### Version

1.7.0

### PR DESCRIPTION
**Pull request by epicfaace**
### Reasons for making this change

Support for allOf -- fixes #52, fixes #1237, fixes #1378

Todo:
- Write tests
- Update docs
- Let [this PR](https://github.com/mokkabonna/json-schema-merge-allof/pull/5) be merged
- Should we depend on json-schema-merge-allof or implement this logic into rjsf directly?

### Checklist

* [ ] **I'm updating documentation**
  - [ ] I've [checked the rendering](https://react-jsonschema-form.readthedocs.io/en/latest/#contributing) of the Markdown text I've added
* [x] **I'm adding or updating code**
  - [ ] I've added and/or updated tests
  - [ ] I've updated [docs](https://react-jsonschema-form.readthedocs.io/) if needed
* [ ] **I'm adding a new feature**
  - [ ] I've updated the playground with an example use of the feature

### CODE REVIEW

**Comment by epicfaace** [comment]
[Here's a sample playground link.](https://deploy-preview-1380--trusting-agnesi-d3347d.netlify.com/#eyJmb3JtRGF0YSI6eyJwYXJ0aWNpcGFudHMiOlsiYiJdfSwic2NoZW1hIjp7ImRlZmluaXRpb25zIjp7InBhcnRpY2lwYW50cyI6eyJ0eXBlIjoiYXJyYXkiLCJpdGVtcyI6eyJ0eXBlIjoic3RyaW5nIn19fSwidHlwZSI6Im9iamVjdCIsInByb3BlcnRpZXMiOnsicGFydGljaXBhbnRzIjp7ImFsbE9mIjpbeyIkcmVmIjoiIy9kZWZpbml0aW9ucy9wYXJ0aWNpcGFudHMifSx7Im1pbkl0ZW1zIjozfV19fX0sInVpU2NoZW1hIjp7fX0=)

**Comment by epicfaace** [comment]
Bug: allOf does not work when one of the options is an array with `uniqueItems: true`, a `minItems` set to greater than 0, and with the checkbox widget.

[Playground reproduction here](https://deploy-preview-1380--trusting-agnesi-d3347d.netlify.com/#eyJmb3JtRGF0YSI6eyJoYXNQYXJ0aWNpcGFudHMiOnRydWV9LCJzY2hlbWEiOnsidHlwZSI6Im9iamVjdCIsInByb3BlcnRpZXMiOnsiaGFzUGFydGljaXBhbnRzIjp7InR5cGUiOiJib29sZWFuIn19LCJkZXBlbmRlbmNpZXMiOnsiaGFzUGFydGljaXBhbnRzIjp7Im9uZU9mIjpbeyJwcm9wZXJ0aWVzIjp7Imhhc1BhcnRpY2lwYW50cyI6eyJjb25zdCI6dHJ1ZX0sInBhcnRpY2lwYW50cyI6eyJhbGxPZiI6W3sidHlwZSI6ImFycmF5IiwiaXRlbXMiOnsidHlwZSI6InN0cmluZyIsImVudW0iOlsiYSIsImIiLCJjIl19LCJ1bmlxdWVJdGVtcyI6dHJ1ZSwibWluSXRlbXMiOjJ9XX19fV19fX0sInVpU2NoZW1hIjp7InBhcnRpY2lwYW50cyI6eyJ1aTp3aWRnZXQiOiJjaGVja2JveGVzIn19fQ==)

**Comment by epicfaace** [comment]
> Bug: When using allOf, with one of the options is an array with `uniqueItems: true`, a `minItems` set to greater than 0, with the checkbox widget, and with omitExtraData and liveOmit set to true, then the user cannot check any checkboxes.
> 
> [Playground reproduction here](https://deploy-preview-1380--trusting-agnesi-d3347d.netlify.com/#eyJmb3JtRGF0YSI6eyJoYXNQYXJ0aWNpcGFudHMiOnRydWV9LCJzY2hlbWEiOnsidHlwZSI6Im9iamVjdCIsInByb3BlcnRpZXMiOnsiaGFzUGFydGljaXBhbnRzIjp7InR5cGUiOiJib29sZWFuIn19LCJkZXBlbmRlbmNpZXMiOnsiaGFzUGFydGljaXBhbnRzIjp7Im9uZU9mIjpbeyJwcm9wZXJ0aWVzIjp7Imhhc1BhcnRpY2lwYW50cyI6eyJjb25zdCI6dHJ1ZX0sInBhcnRpY2lwYW50cyI6eyJhbGxPZiI6W3sidHlwZSI6ImFycmF5IiwiaXRlbXMiOnsidHlwZSI6InN0cmluZyIsImVudW0iOlsiYSIsImIiLCJjIl19LCJ1bmlxdWVJdGVtcyI6dHJ1ZSwibWluSXRlbXMiOjJ9XX19fV19fX0sInVpU2NoZW1hIjp7InBhcnRpY2lwYW50cyI6eyJ1aTp3aWRnZXQiOiJjaGVja2JveGVzIn19fQ==) (check the boxes "Omit extra data" and "Live omit")

This bug was fixed in 4eccffa.

File: package.json [hunk 1]
```diff
@@ -46,6 +46,7 @@
     "@babel/runtime-corejs2": "^7.4.5",
     "ajv": "^6.7.0",
     "core-js": "^2.5.7",
+    "json-schema-merge-allof": "git+https://github.com/mokkabonna/json-schema-merge-allof.git#e43d0f2227476e069667358b86d0e8c91b50a777",
```
Comment (LucianBuzzo) [inline package.json]: Why is this a git dependency?

Comment (epicfaace) [reply package.json]: It links to my PR here (https://github.com/mokkabonna/json-schema-merge-allof/pull/5), which makes sure that mergeAllOf can be called with `deep: false` and thus only merges a top-level `allOf`. It's not strictly required, but I guess it's a performance improvement.

**Comment by erunion** [comment]
Aside from tests and docs, is this feature complete as-is? If not, how can I help out?

**Comment by epicfaace** [comment]
@erunion tests, docs, and a playground example. Feel free to help out with either -- just let me know on this PR!

**Comment by erunion** [comment]
@epicfaace I can help out with a playground example.

**Comment by lorthirk** [comment]
Hello all,

I would be really interested in having this merged. Is there anything I can do to help knowing little to nothing about React?

**Comment by epicfaace** [comment]
Remaining todos:
- update docs
- figure out whether we're going to depend on a fork of json-schema-merge-allof (https://github.com/mokkabonna/json-schema-merge-allof/pull/5) or implement this logic within rjsf ourselves

File: package.json (hunk 1 above)
Comment (epicfaace) [reply package.json]: Since it's not strictly required, I'll just remove it since that's better than depending on a git dependency.

**Comment by PaulNendick** [comment]
Hello All, this is holding up an important fix very far downstream elsewhere - is there any hope of seeing this PR resolved any time soon? Is there anything a novice can do to help? Thanks!

**Comment by PaulNendick** [comment]
Hello All, this is holding up an important fix very far downstream elsewhere - is there any hope of seeing this PR resolved any time soon? Is there anything a novice can do to help? Thanks!

**Comment by edi9999** [comment]
Looks good to me, we can merge if you think it is good

**Comment by PaulNendick** [comment]
Thanks everyone!

**Comment by lorthirk** [comment]
@epicfaace could you please fix the conflicts? We're so close!

**Comment by jorisw** [comment]
By build is currently breaking because of this:

```
node_modules/@rjsf/core/dist/es/utils.js:20
Attempted import error: 'json-schema-merge-allof' does not contain a default export (imported as 'mergeAllOf').
```

@rjsf/core 2.3.0
json-schema-merge-allof 0.5.7 - 0.70.0

**Pull request merged**